presto-spark-base/src/main/java/com/facebook/presto/spark/planner/optimizers/PickJoinSides.java

```diff
@@ -0,0 +1,117 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spark.planner.optimizers;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.cost.StatsProvider;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+
+import static com.facebook.presto.SystemSessionProperties.isSizeBasedJoinDistributionTypeEnabled;
+import static com.facebook.presto.spark.PrestoSparkSessionProperties.isAdaptivePickJoinSidesEnabled;
+import static com.facebook.presto.sql.planner.iterative.rule.DetermineJoinDistributionType.isMuchSmaller;
+import static com.facebook.presto.sql.planner.iterative.rule.DetermineJoinDistributionType.isSmall;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.LEFT;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+
+/**
+ * This optimizer chooses the build and probe side of the join based on the size of the
+ * join inputs, putting the smaller table on the right side/ build side of the join.
+ * <p>
+ * for example, consider a join with two children, nodeA and nodeB
+ * if  nodeA is smaller than nodeB, we will convert
+ *         JOIN                       JOIN
+ *        /    \          = >        /    \
+ *    nodeA   nodeB               nodeB   nodeA
+ */
+public class PickJoinSides
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join().matching(joinNode ->
+            joinNode.getDistributionType().isPresent()
+                    && joinNode.getDistributionType().get() == PARTITIONED
+                    // Flipping right/left non-equality joins is only supported when
+                    // changing the distribution type too
+                    && !(joinNode.getCriteria().isEmpty() && (joinNode.getType() == LEFT || joinNode.getType() == RIGHT)));
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isAdaptivePickJoinSidesEnabled(session);
+    }
+
+    @Override
+    public Result apply(JoinNode joinNode, Captures captures, Context context)
+    {
+        StatsProvider statsProvider = context.getStatsProvider();
+        double leftSize = statsProvider.getStats(joinNode.getLeft()).getOutputSizeInBytes();
+        double rightSize = statsProvider.getStats(joinNode.getRight()).getOutputSizeInBytes();
+
+        if (rightSize > leftSize) {
```

Comment:
Should we have some threshold like we will switch sides if right is x% greater than left ? like muchSmaller ?

Reply:
Will there be any penalty to switch sides? It might be more useful then

Reply:
there's no penalty for switching, so we may as well. If we see problems empirically, we can figure out what the best fix is (including potentially adding some threshold).